infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/env/EnvironmentVariableSorterTest.java

```diff
@@ -0,0 +1,131 @@
+    // breaking given-when-then flow of our tests on purpose here, because the below would result
+    // in 24 test cases I refuse to write out manually
+
+    testSingleValue("value", emptySet(), "no refs");
+    testSingleValue("$(NO_REF", emptySet(), "unclosed ref");
+    testSingleValue("$$(NO_REF)", emptySet(), "escaped ref");
+    testSingleValue("$NO_REF)", emptySet(), "invalid start ref");
+    testSingleValue("$(NO REF)", emptySet(), "invalid name ref");
+    testSingleValue("$(REF)", singleton("REF"), "valid ref");
+  }
+
+  @Test
+  public void shouldSortByNameIfNoDependencies() throws Exception {
+    // given
+    Map<String, EnvVar> vars = vars(var("b", "b"), var("a", "a"), var("c", "c"));
+
+    // when
+    List<EnvVar> sorted = EnvironmentVariablesSorter.sort(vars);
+
+    // then
+    assertEquals(sorted, Arrays.asList(var("a", "a"), var("b", "b"), var("c", "c")));
+  }
+
+  @Test
+  public void shouldSortDependenciesBeforeDependents() throws Exception {
+    // given
+    Map<String, EnvVar> vars = vars(var("a", "$(b)"), var("b", "b"), var("c", "c"));
+
+    // when
+    List<EnvVar> sorted = EnvironmentVariablesSorter.sort(vars);
+
+    // then
+    assertEquals(sorted, Arrays.asList(var("b", "b"), var("c", "c"), var("a", "$(b)")));
+  }
+
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void shouldDetectSelfReference() throws Exception {
+    // given
+    Map<String, EnvVar> vars = vars(var("a", "$(a)"));
+
+    // when
+    EnvironmentVariablesSorter.sort(vars);
+
+    // then the exception is expected
+  }
+
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void shouldDetectDirectCycles() throws Exception {
+    // given
+    Map<String, EnvVar> vars = vars(var("a", "$(b)"), var("b", "$(a)"));
+
+    // when
+    EnvironmentVariablesSorter.sort(vars);
+
+    // then the exception is expected
+  }
+
+  @Test(expectedExceptions = InfrastructureException.class)
+  public void shouldDetectIndirectCycles() throws Exception {
+    // given
+    Map<String, EnvVar> vars = vars(var("a", "$(b)"), var("b", "$(c)"), var("c", "$(a)"));
+
+    // when
+    EnvironmentVariablesSorter.sort(vars);
+
+    // then the exception is expected
+  }
+
+  @Test
+  public void shouldDetectMultipleReferences() throws Exception {
+    // given
+    EnvVar envVar = var("a", "$(b) $(c) $$(d)");
+
+    // when
+    SortedSet<String> refs = EnvironmentVariablesSorter.findReferences(envVar);
+
+    // then
+    assertEquals(refs, new HashSet<>(Arrays.asList("b", "c")));
+  }
+
+  private static EnvVar var(String name, String value) {
+    return new EnvVar(name, value, null);
+  }
+
+  private static Map<String, EnvVar> vars(EnvVar... vars) {
+    return Stream.of(vars).collect(Collectors.toMap(EnvVar::getName, identity()));
+  }
+
+  private static void testSingleValue(String value, Set<String> expected, String caseName) {
+    assertEquals(findReferences(var("name", value)), expected, caseName + ": sole");
```

Comment:
TestNG has very useful feature for such cases - data provider.

Reply:
A data provider could be used if I inverted the logic somewhat and supplied the individual test cases from https://github.com/eclipse/che/pull/12610/files/5b49b89ee272db490296942e2af6e68d9ae0e84e#diff-fe28a35161f1e8e64261b7c68612f0beR35 to a single test that would do the same thing as `testSingleValue`.

IMHO supplying the testcases as Object[][] would actually hurt the readability of the testcase, but that's just my taste. I am happy to rewrite it if you insist.

Reply:
I do not insist! Just wondered whether you know about this capability.
I found reading this method hard. For example, still don't understand what is `": sole"`. Could you explain that to me?

Reply:
sole as in solitary, meaning that the value is taken as is, unlike the rest of the tests where the value is (pre|ap)pended to. I'm modifying the "case name" so that the error message from testng mentions the concrete case that failed.